### PR TITLE
Always use standard KVP.

### PR DESCRIPTION
**Pull request by CarsonBanov**
For: https://github.com/prebid/Prebid.js/issues/196

/cc @mkendall07

### CODE REVIEW

**Comment by mkendall07** [comment]
@CarsonBanov 
Looks good, it would be nice if we issued a warning if the value is being overwritten (meaning it was set previously). Do you think you can add that?

**Comment by CarsonBanov** [comment]
@mkendall07 Added.  I used `utils.logMessage` since the only alternatives were `logError` and `logInfo`.  Can change it to either of those (there's no logWarning right?) if you prefer.

**Comment by mkendall07** [comment]
@CarsonBanov 
I'm actually having trouble replicating the issue for validating the fix. I'm seeing bid keys being set correctly. Can you post a simple test that replicates the behavior?

**Comment by CarsonBanov** [comment]
Try this:
* In your example, what happens if you have bidder 2 win?
* The custom `sendAlways` key will be set, but `hb_pb` will never be set!  This is critical since the prebid line items are set to fire based on `hb_pb`, so `hb_pb` should always be set, right?

To show this more clearly: 
* Set `alwaysUseBid=true` for _both_ 1 and 2 with custom KVP `sendAlways1` and `sendAlways2`
* Will `hb_pb` ever be set?  As far as I can tell, this is no because they will both set the custom KVP but won't set `hb_pb`.

Now we can try to fix this by including `hb_pb` in the custom KVP:
* Set `alwaysUseBid=true` for _both_ 1 and 2.  Bidder 1 has custom KVP `sendAlways1`and `hb_pb`. Bidder 2 has `sendAlways2` and `hb_pb`.
* Have bidder 2 win and then bidder 1 comes in.
* Will `hb_pb` be the correct value?  As far as I can tell, this is no because they will both set their custom KVP, but the value for `hb_pb` will be from bidder 1 overwriting the higher value from bidder 2.

Does that make sense? I hope I am being clear enough.
* The standard KVP `hb_pb` should always hold the winning bid independent of `alwaysUseBid`
* The custom KVP should be sent if the bidder wins, or if `alwaysUseBid=true` (but `hb_pb` should still hold the winning bid)

**Comment by CarsonBanov** [comment]
.cc @mkendall07

**Comment by mkendall07** [comment]
@CarsonBanov Thanks - very helpful. Will try those scenarios out now.

**Comment by mkendall07** [comment]
I think the assumption was that if your using `alwaysSendBid` your keys will be sent the way you want. It looks like the side effect of that is that those bids get knocked out of the auction, which doesn't feel right to me. Even with this PR that's not working correctly. I can fix it by including those bids in the auction and that seems to work. I'm going to spend more time on this item because it's so core to how prebid.js works. At one point we had test cases for this but it looks like we might have lost them in a refactor. I'm going to add them back...

**Comment by CarsonBanov** [comment]
@mkendall07 Thanks so much!

Yeah, I'm realizing now that you mention it that this PR will include the standard `hb_` KVPs along with the bid, but since that bid will in some cases be marked as `alwaysUseBid` it will get dropped out of `buildBidResponse`.  I just added a commit to this PR that you can look at to see if it helps.

**Comment by mkendall07** [comment]
@CarsonBanov 
Does adding those bids into the auction solve your use case without the KV changes you made? I just committed the old tests, and 2 are failing with your changes:
https://github.com/prebid/Prebid.js/commit/1aa69d25e9f15d31b89c7d7ba28c9c78a6202149

**Comment by CarsonBanov** [comment]
@mkendall07 

I think that would be better.  The custom KVP will be set no matter what, but the bid would still be included in the auction.  Unfortunately, the "winner of the auction" will set the KVP from the bidder, so if the custom bidder wins, it will just re-set the custom KVP (and never set the `hb_pb` KVPs).
I think that means in all cases where we use `alwaysUseBid ` we would need to include all of the `hb_pb` KVP as well since it won't inherit from standard.

For example, this config setup _will never set `hb_pb` KVPs_:
```
pbjs.bidderSettings =
    standard:
      adserverTargeting: [
          key: 'hb_pb'
          val: (bidResponse) ->
            prebidRoundFunction bidResponse.cpm
      ]
    # Custom brealtime
    brealtime:
      alwaysUseBid: true
      adserverTargeting: [
          key: 'custom_brt'
          val: (bidResponse) ->
            bidResponse.bidder
      ]
    # Custom appnexus
    appnexus:
      alwaysUseBid: true
      adserverTargeting: [
          key: 'custom_appnexus'
          val: (bidResponse) ->
            bidResponse.bidder
      ]
```
Instead, we would have to do something like this (duplicately include the `hb_pb` with all custom targeting):
```
pbjs.bidderSettings =
    standard:
      adserverTargeting: [
          key: 'hb_pb'
          val: (bidResponse) ->
            prebidRoundFunction bidResponse.cpm
      ]
    # Custom brealtime
    brealtime:
      alwaysUseBid: true
      adserverTargeting: [
          key: 'custom_brt'
          val: (bidResponse) ->
            bidResponse.bidder
          ,
          key: 'hb_pb'
          val: (bidResponse) ->
            prebidRoundFunction bidResponse.cpm
      ]
    # Custom appnexus
    appnexus:
      alwaysUseBid: true
      adserverTargeting: [
          key: 'custom_appnexus'
          val: (bidResponse) ->
            bidResponse.bidder
          ,
          key: 'hb_pb'
          val: (bidResponse) ->
            prebidRoundFunction bidResponse.cpm
      ]
```

**Comment by CarsonBanov** [comment]
Another thing is that the `hb_pb` KVP will be set by all of those `alwaysUseBid` partners _every bid_ which means we are relying on the line `pb_targetingMap[adUnitCode] = utils.extend(pb_targetingMap[adUnitCode], keyValues);` to overwrite them for the winner of `var winningBid = getWinningBid(bidArrayTargeting);`

**Comment by mkendall07** [comment]
I see your what your after, I think it's creating other side affects though, like I pointed out above with using default settings instead of what is configured. We'll need to fix this properly to support all use cases.

**Comment by CarsonBanov** [comment]
tl;dr
* Adding `alwaysUseBid` bidders' [bids into the auction](https://github.com/prebid/Prebid.js/pull/262/commits/a12c26c4896830033f4c43c609d92caab3a89123) will help. 
* With that change alone we will need to duplicately include `hb_pb` in all `alwaysUseBid` custom adserverTargeting declarations, and the `hb_pb` will be overwritten many times over the course of the auction because of it.  But overall I this behavior should work for us.
* [Always setting the `hb_` KVPs](https://github.com/prebid/Prebid.js/pull/262/commits/90c0f65723db6fb587b2558ad55bd5d54851a73a) would help more by making the `hb_` KVPs more independent of the custom adserverTargeting.  But this isn't extremely necessary I guess.

**Comment by CarsonBanov** [comment]
@mkendall07 Thanks for the attention.  I guess `alwaysUseBid` is supposed to be more for targeting a different set of line items like you outlined above (always set those custom KVP) rather than my particular use case.  

I wonder if an orthogonal solution would be to add a new feature very similar to `bidCpmAdjustment` called something like `kvpAdjustment` which would let us set a custom KVP per partner totally unrelated to `hb_pb`, custom targeting, and the prebid internal auction.




Anyways, thanks for your help.  Feel free to close this and implement whatever fix you feel fit.

**Comment by mkendall07** [comment]
@CarsonBanov 
I'm wondering now if we switched the order to always apply `default` settings for KVP and then apply the custom settings if that would solve all use cases (along with adding all bids into the auction)... will play with that and create a new PR for you to look at.

**Comment by mkendall07** [comment]
Following up with #295
